Firebase/Storage/FIRStorageUploadTask.m

```diff
@@ -68,145 +70,154 @@ - (void)dealloc {
+        [self removeAllObservers];
+        self->_fetcherCompletion = nil;
+        return;
+      }
+
+      // Upload completed successfully, fire completion callbacks
+      self.state = FIRStorageTaskStateSuccess;
+
+      NSDictionary *responseDictionary = [NSDictionary frs_dictionaryFromJSONData:data];
+      if (responseDictionary) {
+        FIRStorageMetadata *metadata =
+            [[FIRStorageMetadata alloc] initWithDictionary:responseDictionary];
+        [metadata setType:FIRStorageMetadataTypeFile];
+        self.metadata = metadata;
+      } else {
+        self.error = [FIRStorageErrors errorWithInvalidRequest:data];
+      }
+
+      [self fireHandlersForStatus:FIRStorageTaskStatusSuccess snapshot:self.snapshot];
       [self removeAllObservers];
       self->_fetcherCompletion = nil;
-      return;
-    }
-
-    // Upload completed successfully, fire completion callbacks
-    self.state = FIRStorageTaskStateSuccess;
-
-    NSDictionary *responseDictionary = [NSDictionary frs_dictionaryFromJSONData:data];
-    if (responseDictionary) {
-      FIRStorageMetadata *metadata =
-          [[FIRStorageMetadata alloc] initWithDictionary:responseDictionary];
-      [metadata setType:FIRStorageMetadataTypeFile];
-      self.metadata = metadata;
-    } else {
-      self.error = [FIRStorageErrors errorWithInvalidRequest:data];
-    }
-
-    [self fireHandlersForStatus:FIRStorageTaskStatusSuccess snapshot:self.snapshot];
-    [self removeAllObservers];
-    self->_fetcherCompletion = nil;
-  };
+    };
 #pragma clang diagnostic pop
 
-  [_uploadFetcher
-      beginFetchWithCompletionHandler:^(NSData *_Nullable data, NSError *_Nullable error) {
-        weakSelf.fetcherCompletion(data, error);
-      }];
+    [strongSelf->_uploadFetcher
+        beginFetchWithCompletionHandler:^(NSData *_Nullable data, NSError *_Nullable error) {
+          weakSelf.fetcherCompletion(data, error);
+        }];
+  }];
 }
 
 #pragma mark - Upload Management
 
 - (void)cancel {
-  NSAssert([NSThread isMainThread],
-           @"Cancel attempting to execute on non main queue! Please only "
-           @"execute this method on the main queue.");
-  self.state = FIRStorageTaskStateCancelled;
-  [_uploadFetcher stopFetching];
-  if (self.state != FIRStorageTaskStateSuccess) {
-    self.metadata = _uploadMetadata;
-  }
-  self.error = [FIRStorageErrors errorWithCode:FIRStorageErrorCodeCancelled];
-  [self fireHandlersForStatus:FIRStorageTaskStatusFailure snapshot:self.snapshot];
+  __weak FIRStorageUploadTask *weakSelf = self;
+
+  [self dispatchAsync:^() {
+    weakSelf.state = FIRStorageTaskStateCancelled;
+    [weakSelf.uploadFetcher stopFetching];
+    if (weakSelf.state != FIRStorageTaskStateSuccess) {
+      weakSelf.metadata = weakSelf.uploadMetadata;
+    }
+    weakSelf.error = [FIRStorageErrors errorWithCode:FIRStorageErrorCodeCancelled];
+    [weakSelf fireHandlersForStatus:FIRStorageTaskStatusFailure snapshot:weakSelf.snapshot];
```

Comment:
Should this block have a `strongSelf` to guarantee these callbacks are invoked?

Reply:
Same as above :)